### ISSUE TITLE
implement plan

### ISSUE THREAD

**Issue by drzo** [ReZorg/marduk-ml-sdk#10]
<div class="markdown-body MarkdownRenderer-module__container__sJ8jf ExpandableMarkdown-module__content__Rn4oS UserProblemStatementMessage-module__messageText__Jng0H" data-copilot-markdown="true" style="--MarkdownRenderer_streaming-fade-duration: 750ms;"><p>Implement the plan:</p>
<p>Determination: this is a planning/analysis task, not a code-change task.</p>
<h2>Current codebase baseline</h2>
<ul>
<li>The repository is a Cloudflare VibeSDK scaffold rebranded toward Marduk’s Lab.</li>
<li>Existing foundation:
<ul>
<li>React/Vite frontend in <code node="[object Object]">/tmp/workspace/ReZorg/marduk-ml-sdk/src</code>.</li>
<li>Cloudflare Worker backend in <code node="[object Object]">/tmp/workspace/ReZorg/marduk-ml-sdk/worker</code>.</li>
<li>Durable Object code generation agent.</li>
<li>D1 app/user database schema.</li>
<li>R2 template bucket support.</li>
<li>KV bindings for Marduk state.</li>
<li>TypeScript SDK in <code node="[object Object]">/tmp/workspace/ReZorg/marduk-ml-sdk/sdk</code>.</li>
<li>Marduk cognitive SDK primitives under <code node="[object Object]">/tmp/workspace/ReZorg/marduk-ml-sdk/sdk/src/marduk</code>.</li>
<li>Existing <code node="[object Object]">cognitive</code> project type and Mad-Lab ML frontend feature shell.</li>
<li>Existing ML prompt extension and five ML template descriptors.</li>
</ul>
</li>
</ul>
<h2>Implementation plan</h2>
<h3>1. Define the complete Mad-Lab product surface</h3>
<p>Establish Mad-Lab as a first-class autonomous ML workbench on top of the VibeSDK scaffold with these core areas:</p>
<ul>
<li>ML project generation</li>
<li>Dataset management</li>
<li>Experiment tracking</li>
<li>Training orchestration</li>
<li>Model evaluation</li>
<li>Model serving</li>
<li>AutoML / MOSES optimization</li>
<li>Cognitive memory and recall</li>
<li>Archon specialist agents</li>
<li>Autonomy heartbeat and recommendations</li>
<li>GitHub export / deployment</li>
<li>SDK-driven programmatic workflows</li>
</ul>
<h3>2. Expand project types and capabilities</h3>
<p>Extend the current <code node="[object Object]">cognitive</code> feature into a richer ML workbench capability model.</p>
<p>Relevant areas:</p>
<ul>
<li><code node="[object Object]">/tmp/workspace/ReZorg/marduk-ml-sdk/worker/agents/core/types.ts</code></li>
<li><code node="[object Object]">/tmp/workspace/ReZorg/marduk-ml-sdk/worker/agents/core/features/types.ts</code></li>
<li><code node="[object Object]">/tmp/workspace/ReZorg/marduk-ml-sdk/src/features/cognitive</code></li>
<li><code node="[object Object]">/tmp/workspace/ReZorg/marduk-ml-sdk/src/features/index.ts</code></li>
</ul>
<p>Add clear capability flags for:</p>
<ul>
<li>Training jobs</li>
<li>Evaluation runs</li>
<li>Dataset uploads</li>
<li>Model artifact storage</li>
<li>Experiment dashboards</li>
<li>Inference endpoints</li>
<li>AutoML search</li>
<li>Agent-assisted debugging</li>
<li>Cognitive memory recall</li>
</ul>
<h3>3. Turn ML templates into real scaffolded project starters</h3>
<p>The current <code node="[object Object]">ML_TEMPLATE_REGISTRY</code> only describes templates. Convert it into a real template system backed by R2 or bundled descriptors.</p>
<p>Template families:</p>
<ul>
<li>PyTorch training pipeline</li>
<li>FastAPI model serving</li>
<li>Data pipeline</li>
<li>Reinforcement learning environment</li>
<li>HuggingFace fine-tuning</li>
<li>RAG / embedding pipeline</li>
<li>AutoML experiment</li>
<li>Evaluation benchmark suite</li>
</ul>
<p>Each template should define:</p>
<ul>
<li>File tree</li>
<li>Required dependencies</li>
<li>Entrypoints</li>
<li>Expected commands</li>
<li>Preview behavior</li>
<li>Training/evaluation lifecycle</li>
<li>Export/deploy targets</li>
</ul>
<h3>4. Add ML workbench domain persistence</h3>
<p>Extend the database layer beyond generated apps.</p>
<p>Add persistence for:</p>
<ul>
<li>Datasets</li>
<li>Dataset versions</li>
<li>Experiments</li>
<li>Runs</li>
<li>Metrics</li>
<li>Artifacts</li>
<li>Models</li>
<li>Model versions</li>
<li>Evaluation reports</li>
<li>Training jobs</li>
<li>Inference deployments</li>
<li>AutoML studies</li>
<li>Cognitive memory links</li>
</ul>
<p>Relevant area:</p>
<ul>
<li><code node="[object Object]">/tmp/workspace/ReZorg/marduk-ml-sdk/worker/database/schema.ts</code></li>
<li><code node="[object Object]">/tmp/workspace/ReZorg/marduk-ml-sdk/migrations</code></li>
</ul>
<p>Use D1 for metadata, R2 for artifacts, KV for fast cognitive/session state, and future Vectorize support for semantic retrieval.</p>
<h3>5. Build ML backend APIs</h3>
<p>Add first-class backend routes/controllers for Mad-Lab workbench operations.</p>
<p>Suggested API groups:</p>
<ul>
<li><code node="[object Object]">/api/ml/datasets</code></li>
<li><code node="[object Object]">/api/ml/experiments</code></li>
<li><code node="[object Object]">/api/ml/runs</code></li>
<li><code node="[object Object]">/api/ml/artifacts</code></li>
<li><code node="[object Object]">/api/ml/models</code></li>
<li><code node="[object Object]">/api/ml/evaluations</code></li>
<li><code node="[object Object]">/api/ml/jobs</code></li>
<li><code node="[object Object]">/api/ml/automl</code></li>
<li><code node="[object Object]">/api/ml/agents</code></li>
<li><code node="[object Object]">/api/ml/memory</code></li>
<li><code node="[object Object]">/api/ml/autonomy</code></li>
</ul>
<p>Relevant areas:</p>
<ul>
<li><code node="[object Object]">/tmp/workspace/ReZorg/marduk-ml-sdk/worker/api/routes</code></li>
<li><code node="[object Object]">/tmp/workspace/ReZorg/marduk-ml-sdk/worker/api/controllers</code></li>
<li><code node="[object Object]">/tmp/workspace/ReZorg/marduk-ml-sdk/worker/services</code></li>
</ul>
<h3>6. Create a real ML execution layer</h3>
<p>Use the existing Cloudflare Container sandbox model for generated app previews and ML execution.</p>
<p>Execution capabilities:</p>
<ul>
<li>Run training scripts</li>
<li>Stream logs</li>
<li>Capture metrics</li>
<li>Store checkpoints</li>
<li>Run evaluation scripts</li>
<li>Export ONNX/model artifacts</li>
<li>Start model-serving previews</li>
<li>Support cancellation and retry</li>
<li>Enforce resource limits</li>
</ul>
<p>Relevant areas:</p>
<ul>
<li><code node="[object Object]">/tmp/workspace/ReZorg/marduk-ml-sdk/worker/services/sandbox</code></li>
<li><code node="[object Object]">/tmp/workspace/ReZorg/marduk-ml-sdk/SandboxDockerfile</code></li>
<li><code node="[object Object]">/tmp/workspace/ReZorg/marduk-ml-sdk/wrangler.jsonc</code></li>
</ul>
<h3>7. Deepen Marduk cognitive integration</h3>
<p>The current worker-side <code node="[object Object]">MardukService</code> is lightweight. Expand it into a real cognitive service layer.</p>
<p>Relevant area:</p>
<ul>
<li><code node="[object Object]">/tmp/workspace/ReZorg/marduk-ml-sdk/worker/services/marduk/MardukService.ts</code></li>
<li><code node="[object Object]">/tmp/workspace/ReZorg/marduk-ml-sdk/sdk/src/marduk</code></li>
</ul>
<p>Add:</p>
<ul>
<li>Persistent declarative memory for project facts</li>
<li>Episodic memory for user interactions and run history</li>
<li>Procedural memory for reusable ML workflows</li>
<li>Semantic memory for datasets, models, metrics, and concepts</li>
<li>Associative recall during blueprint generation and chat</li>
<li>Memory summarization and compaction</li>
<li>Memory-to-prompt context injection</li>
<li>User/project scoped memory isolation</li>
</ul>
<h3>8. Implement autonomous ML agents</h3>
<p>Build on the existing <code node="[object Object]">marduk_archon_build</code> tool and turn Archon into a practical agent factory.</p>
<p>Agent types:</p>
<ul>
<li>Dataset profiler</li>
<li>Feature engineering assistant</li>
<li>Training engineer</li>
<li>Evaluation analyst</li>
<li>Hyperparameter optimizer</li>
<li>Model-serving engineer</li>
<li>Debugging assistant</li>
<li>MLOps release assistant</li>
<li>Research/RAG assistant</li>
</ul>
<p>Each agent should have:</p>
<ul>
<li>Purpose</li>
<li>Capabilities</li>
<li>Tool access</li>
<li>Memory scope</li>
<li>Safety limits</li>
<li>Audit trail</li>
<li>Reuse across sessions</li>
</ul>
<h3>9. Add AutoML and MOSES workflows</h3>
<p>Use the existing MOSES SDK primitives to support evolutionary search.</p>
<p>Capabilities:</p>
<ul>
<li>Hyperparameter search</li>
<li>Multi-objective optimization</li>
<li>Architecture search metadata</li>
<li>Fitness tracking</li>
<li>Trial comparison</li>
<li>Best-run promotion</li>
<li>Auto-generated config updates</li>
<li>Optional PR creation/export workflow</li>
</ul>
<p>Relevant area:</p>
<ul>
<li><code node="[object Object]">/tmp/workspace/ReZorg/marduk-ml-sdk/sdk/src/marduk/moses</code></li>
<li><code node="[object Object]">/tmp/workspace/ReZorg/marduk-ml-sdk/worker/services/marduk</code></li>
<li>GitHub workflow files if CI-based evolution is used</li>
</ul>
<h3>10. Build the Mad-Lab frontend workbench</h3>
<p>Expand <code node="[object Object]">/tmp/workspace/ReZorg/marduk-ml-sdk/src/features/cognitive</code> from a preview shell into a full workbench.</p>
<p>Views:</p>
<ul>
<li>Overview</li>
<li>Code</li>
<li>Preview</li>
<li>Terminal</li>
<li>Blueprint</li>
<li>Datasets</li>
<li>Experiments</li>
<li>Runs</li>
<li>Metrics</li>
<li>Models</li>
<li>Evaluation</li>
<li>Serving</li>
<li>Agents</li>
<li>Memory</li>
<li>Autonomy</li>
</ul>
<p>UI functionality:</p>
<ul>
<li>Dataset upload/import panel</li>
<li>Experiment dashboard</li>
<li>Training run timeline</li>
<li>Live log terminal</li>
<li>Metrics charts</li>
<li>Confusion matrix / evaluation cards</li>
<li>Model artifact browser</li>
<li>Inference tester</li>
<li>AutoML study viewer</li>
<li>Agent builder panel</li>
<li>Cognitive memory explorer</li>
<li>Autonomy recommendation inbox</li>
</ul>
<h3>11. Extend the SDK for Mad-Lab workflows</h3>
<p>Add programmatic ML workbench APIs to the SDK.</p>
<p>Relevant area:</p>
<ul>
<li><code node="[object Object]">/tmp/workspace/ReZorg/marduk-ml-sdk/sdk/src</code></li>
</ul>
<p>SDK capabilities:</p>
<ul>
<li>Create ML project</li>
<li>Select ML template</li>
<li>Upload/register dataset</li>
<li>Start training</li>
<li>Stream logs</li>
<li>Read metrics</li>
<li>Compare runs</li>
<li>Trigger evaluation</li>
<li>Export model</li>
<li>Deploy serving endpoint</li>
<li>Run AutoML search</li>
<li>Query cognitive memory</li>
<li>Build Archon agent</li>
</ul>
<h3>12. Improve generation prompts for ML projects</h3>
<p>The existing ML prompt extension is a strong start. Extend it into template-aware, lifecycle-aware generation.</p>
<p>Relevant area:</p>
<ul>
<li><code node="[object Object]">/tmp/workspace/ReZorg/marduk-ml-sdk/worker/agents/prompts.ts</code></li>
<li><code node="[object Object]">/tmp/workspace/ReZorg/marduk-ml-sdk/worker/agents/planning/blueprint.ts</code></li>
</ul>
<p>Add prompt context for:</p>
<ul>
<li>Selected ML template</li>
<li>Dataset availability</li>
<li>Experiment target metric</li>
<li>Runtime constraints</li>
<li>Deployment target</li>
<li>Reproducibility requirements</li>
<li>Evaluation criteria</li>
<li>Existing memory recall</li>
<li>Suggested Archon agents</li>
</ul>
<h3>13. Add security and governance controls</h3>
<p>Add workbench-specific protections:</p>
<ul>
<li>Per-user project isolation</li>
<li>Dataset access controls</li>
<li>Artifact permissions</li>
<li>Signed artifact URLs</li>
<li>Job resource limits</li>
<li>Command allowlists</li>
<li>Secret redaction</li>
<li>Model-serving abuse limits</li>
<li>API key scope enforcement</li>
<li>Audit logs for agent actions</li>
</ul>
<h3>14. Add observability and autonomy</h3>
<p>Expand the existing scheduled heartbeat into a useful autonomous operations loop.</p>
<p>Current area:</p>
<ul>
<li><code node="[object Object]">/tmp/workspace/ReZorg/marduk-ml-sdk/worker/index.ts</code></li>
<li><code node="[object Object]">/tmp/workspace/ReZorg/marduk-ml-sdk/worker/services/marduk/MardukService.ts</code></li>
</ul>
<p>Autonomy should monitor:</p>
<ul>
<li>Failed training jobs</li>
<li>Memory saturation</li>
<li>Cost/usage trends</li>
<li>Slow templates</li>
<li>Broken deployments</li>
<li>Repeated user corrections</li>
<li>Model quality regressions</li>
<li>Sandbox failures</li>
<li>Stale experiments</li>
</ul>
<p>It should produce:</p>
<ul>
<li>Health reports</li>
<li>Optimization suggestions</li>
<li>Suggested template fixes</li>
<li>Suggested AutoML runs</li>
<li>Suggested agent creation</li>
<li>Suggested documentation updates</li>
</ul>
<h3>15. Add testing and validation strategy</h3>
<p>Use existing repo commands:</p>
<ul>
<li><code node="[object Object]">npm run typecheck</code></li>
<li><code node="[object Object]">npm run lint</code></li>
<li><code node="[object Object]">npm run test</code></li>
<li><code node="[object Object]">npm run build</code></li>
</ul>
<p>Add targeted coverage for:</p>
<ul>
<li>ML template selection</li>
<li>Capabilities endpoint</li>
<li>Cognitive project creation</li>
<li>Marduk memory persistence</li>
<li>Archon agent creation</li>
<li>Training job lifecycle</li>
<li>Experiment/run APIs</li>
<li>SDK ML workflows</li>
<li>Frontend cognitive views</li>
<li>Autonomy heartbeat</li>
</ul>
<h3>16. Suggested implementation order</h3>
<ol>
<li>Solidify data model and API contracts.</li>
<li>Make ML templates real and selectable.</li>
<li>Expand the <code node="[object Object]">cognitive</code> frontend into a workbench shell.</li>
<li>Add experiment/run/model persistence.</li>
<li>Add job execution and log streaming.</li>
<li>Add metrics/artifact capture.</li>
<li>Add SDK ML APIs.</li>
<li>Add cognitive memory recall into generation/chat.</li>
<li>Add Archon specialist agents.</li>
<li>Add AutoML/MOSES workflows.</li>
<li>Add autonomy dashboard and recommendations.</li>
<li>Harden security, quotas, observability, and tests.</li>
</ol></div>
